### PR TITLE
Support user-level workspace config keyed by Git remote

### LINKED ISSUE #13815: Support user-level workspace config keyed by Git remote
## Problem

Workspace configuration is currently repository-local. Users need a way to keep personal workspace overrides outside the repository so they can customize a workspace without changing its checked-in dagger.toml.

A user-level Dagger config should be able to select overrides by the workspace's Git remote. This should include adding personal environments.

## Desired behavior

- Load workspace overrides from a documented user-level config file.
- Key workspace-specific overrides by a normalized Git remote.
- Merge the matching user-level overrides over the repository's workspace config without modifying the repository.
- Allow user-level overrides to add environments.
- Define clear behavior for workspaces with no remote, multiple remotes, and equivalent remote URL forms.

The exact config path and schema can follow the current workspace configuration design.

## Prior work

Draft implementation and design exploration: #13265. That PR used ~/.config/dagger/config.toml, derived workspace keys from Git origins, and added user-level environments; it was later closed as stale.

### PR DESCRIPTION
**Pull request by kpenfound**
Closes #13815

## Summary

Adds user-level workspace configuration: personal overrides kept outside the repository in the user's Dagger config file (`~/.config/dagger/config.toml`, or `$DAGGER_CONFIG` — the same file that already holds `[llm]`), keyed by the workspace's normalized Git remote.

```toml
# Always applied when working in the github.com/acme/api workspace:
[workspaces."github.com/acme/api".modules.aws.settings]
profile = "alice-dev"

# A personal environment, selected with `dagger --env dev ...`:
[workspaces."github.com/acme/api".env.dev.modules.aws.settings]
region = "us-west-2"
```

- User-level values merge over the repository's `dagger.toml` without modifying it; user-level environments are added to (and merge over) the repository's own. Merge order is base → user overlay → selected `--env` overlay, applied consistently across module loading, `workspace config` reads, `envList`, and `dagger settings` display.
- The workspace key is the normalized `origin` remote for local checkouts (worktree/submodule `.git` files and `commondir` indirection included) and the clone address for `-W` remote workspaces. Equivalent remote spellings (`git@github.com:acme/api.git`, `https://github.com/acme/api`, `github.com/acme/api`) all match, on both sides of the comparison.
- `-g/--global` on `dagger settings` and `dagger workspace config` stores/unsets values user-level instead of in the repository, composing with `--env`. Writes happen client-side under the shared config file's lock, preserving unrelated sections. `-g` on a read errors: reads always show the effective merged view.

## Defined edge behavior (each pinned by a test)

- No remote, or a purely local/filesystem remote → no key, no overrides.
- Multiple remotes → keyed by `origin` only (fork checkouts key by the fork's URL).
- Missing user config file, or clients that don't declare a config path (older CLIs) → unchanged behavior; malformed user config errors loudly.
- An always-on user entry naming a module that doesn't exist in the current checkout is skipped, not an error — one key spans every branch and clone of a repo. User-level *environments* stay strictly validated when selected with `--env`.
- Only `modules.<name>.settings.*` (optionally under `env.<name>.*`) can be stored user-level.

## Implementation notes

- `core/workspace/userconfig.go`: parsing, `NormalizeGitRemote`, overlay matching, merge, and format-preserving write/delete helpers.
- New `ClientMetadata.UserConfigPath` carries the caller-host path (CLI fills it from the existing `llmconfig` path resolution); the engine reads the file through the caller host session during workspace load and stores the matched overlay in private (non-GraphQL) fields on `core.Workspace`.
- Docs: new "User-level configuration" section in the workspace configuration reference, a "Personal overrides" section in the environments guide, and a regenerated CLI reference.

Design follows the direction from the closed draft #13265 and its design doc, updated for the current CLI (no `dagger env` lifecycle commands; workspace-key-first schema so always-on overrides and personal envs both work).

## Validation

All test-driven, no host state involved (fake filesystems and `t.TempDir()` + `$DAGGER_CONFIG`):

- `core/workspace`: new unit tests for parsing, normalization, matching, merge, git-config parsing, and user-config write/delete.
- `engine/server`: 12 new session-level tests (fake host FS); full package passes (128).
- `core/schema`: effective-config and field-privacy tests; full package passes (158).
- `core/integration`: new `TestWorkspaceUserConfig` + `TestWorkspaceUserConfigWrites` (14 subtests, real CLI against dev engine); pre-existing settings/env/config suites pass as regression guard (79), plus the full CLI package tests (297).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: core/workspace/userconfig.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+// normalize to "".
+func NormalizeGitRemote(remote string) string {
+	remote = strings.TrimSpace(remote)
+	if remote == "" {
+		return ""
+	}
+	// Filesystem remotes (path or file:// forms) are not usable as keys.
+	if strings.HasPrefix(remote, "/") || strings.HasPrefix(remote, ".") ||
+		strings.HasPrefix(remote, "file://") || strings.HasPrefix(remote, "~") {
+		return ""
```
Comment (grouville) [inline core/workspace/userconfig.go]: Looks like Dagger accepts Windows-local origins when writing user-level workspace config with `--global`.

```text
local Windows path → non-empty workspace key → global write succeeds
```

#### Repro

For example, from an existing Dagger workspace:

```powershell
git remote set-url origin 'C:\Users\alice\api'
dagger workspace config --global modules.aws.settings.profile alice-dev
```

If the repository does not have an origin yet:

```powershell
git remote add origin 'C:\Users\alice\api'
dagger workspace config --global modules.aws.settings.profile alice-dev
```

The Dagger command currently succeeds and creates a user-level workspace entry keyed by `C:\Users\alice\api`. I think it should fail instead: this origin only identifies a path on my machine, just like a Linux-local origin such as `/home/alice/api`. The same issue applies to UNC origins such as `\\server\share\api`.

#### Solution 

We already have `pathutil.GetDrive` for Windows drive letters and UNC shares. Could we normalize the separators and reuse it here so these paths return no workspace key?

I added a small regression test covering drive and UNC paths here: [grouville/dagger@b0ef1aaa9](https://github.com/grouville/dagger/commit/b0ef1aaa9a53a84a856817ae82283cfc0168bb50).

File: engine/server/session_workspaces.go [hunk 2]
```diff
@@ -876,6 +900,109 @@ func (srv *Server) buildCoreWorkspace(
...
+
+// localWorkspaceUserConfigKey derives the user-config key for a local
+// workspace from its git origin remote. Best-effort: a workspace without a
+// usable origin has no key and matches no user-level overrides.
+func localWorkspaceUserConfigKey(
+	ctx context.Context,
+	readFile func(context.Context, string) ([]byte, error),
+	root string,
+) string {
+	data, err := readLocalGitConfig(ctx, readFile, root)
```
Comment (grouville) [inline engine/server/session_workspaces.go]: Looks like user-level workspace config writes and reads disagree when the origin comes from an included Git config.

```text
write: git config --get → include resolved → workspace key found
read:  raw .git/config   → include ignored  → no workspace key
```

For example, from an existing Dagger workspace whose main `.git/config` does not define an origin:

```shell
git config --file .git/origin.config remote.origin.url git@github.com:acme/api.git
git config --local include.path origin.config

dagger workspace config --global modules.aws.settings.profile alice-dev
dagger workspace config modules.aws.settings.profile
```

The global write succeeds because the CLI uses `git config --get`, which finds the included origin. The following read does not apply `alice-dev`, because the engine only parses the main `.git/config` bytes and never sees that origin.

I think both sides should resolve `remote.origin.url` the same way. Could the engine follow Git includes here, or receive the origin already resolved by the client, before normalizing the workspace key?

I added a small regression test covering both `[include]` and `[includeIf]` here: [grouville/dagger@b0ef1aaa9](https://github.com/grouville/dagger/commit/b0ef1aaa9a53a84a856817ae82283cfc0168bb50).

File: core/workspace/userconfig.go (hunk 1 above)
Comment (kpenfound) [reply core/workspace/userconfig.go]: Good catch — fixed in f272b86f3, and I cherry-picked your regression tests as-is (b0ef1aaa9).

`NormalizeGitRemote` now normalizes separators and rejects anything `pathutil.GetDrive` recognizes, alongside the existing POSIX/file:// checks — so `C:\Users\alice\api`, `C:/Users/alice/api`, and `\\server\share\api` all produce no workspace key, which also makes the `--global` write fail with the "not a usable git remote" error like a Linux-local origin does. (`\\?\C:\...` long-path forms fall out of the same check.)

File: engine/server/session_workspaces.go (hunk 2 above)
Comment (kpenfound) [reply engine/server/session_workspaces.go]: Fixed in f272b86f3 (your regression test cherry-picked in b0ef1aaa9) — went with the "engine follows includes" option so the engine and `git config --get` see the same effective config regardless of which side computes the key.

`workspace.ResolveGitConfigIncludes` inlines `include`/`includeIf` content at the directive's location like git does, recursively with git's depth limit of 10. `gitdir:`/`gitdir/i:` conditions match against the per-worktree gitdir with git's pattern transformations (`./` anchored to the config file's dir, `**/` prepended for unanchored patterns, trailing `/` → `**`), and `onbranch:` matches the branch parsed from HEAD. Two documented limitations, both treated as non-matching: `~/`-relative patterns (the caller's home isn't resolvable engine-side) and `hasconfig` conditions. `GitRemoteURL` also now takes the *last* url value like `git config --get`, so an include that overrides the origin wins on both sides.

Unit coverage for the matcher edge cases is in `core/workspace/gitconfig_test.go`, plus engine-level tests for non-matching `gitdir` and `onbranch` on/off the named branch.

**Comment by kpenfound** [comment]
Rebased to fix the DCO check: the cherry-picked regression-test commit had no author sign-off, which I couldn't add on @grouville's behalf, so the tests are now committed under my authorship with `Co-authored-by` credit (0c646d5fd), with the fix re-applied on top (a9039f53e). The tree is unchanged from the previously pushed state.

**Comment by grouville** [comment]
Quick UX question: should this work when `-W` points to a remote workspace?

```console
dagger -W github.com/acme/ws workspace config -g KEY VALUE
```

`-g` only writes to the user's local config, keyed by the remote. It does not change the remote workspace. We already derive the key for remote workspaces, and _**global unsets are allowed today**_.

I think global sets should work too, for consistency. Does that match the intended UX? If so, there is still a gap in the current implementation: the local-only workspace policy rejects this command before the global-write handler runs.

**Review by grouville — APPROVED** [review]
Left a small question on the UX regarding _setting global configs_ with remote workspaces. Apart from that, LGTM ✅

**Comment by kpenfound** [comment]
Added support for `--global` writes against remote workspaces (9b445a7c8): `dagger -W https://github.com/acme/api settings -g <module> <key> <value>` now works — the remote workspace is read for the key and settings introspection, and the write lands in the caller's local user config keyed by the clone address.

Two gaps were fixing this: the CLI's workspace flag policy forced `-W` to be a local path for settings/config *writes* (right for repo-config writes, wrong for `-g`, which is always caller-local — now exempted), and the engine's late-metadata merge in `getOrInitClient` didn't carry `UserConfigPath` onto clients created by an earlier request, so nested sessions loaded workspaces without user-config awareness. Covered by CLI unit tests for remote-ref key derivation/policy plus an integration test driving both commands against a served git remote.
